File: typescript/problem_44.ts
export function countInversions(arr: number[]): number {
    function mergeSort(a: number[]): [number[], number] {
        if (a.length <= 1) return [a, 0];
        const mid = a.length >> 1;
        const [left, lc] = mergeSort(a.slice(0, mid));
        const [right, rc] = mergeSort(a.slice(mid));
        const merged: number[] = [];
        let i = 0, j = 0, mc = 0;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) merged.push(left[i++]);
            else { merged.push(right[j++]); mc += left.length - i; }
        }
        return [[...merged, ...left.slice(i), ...right.slice(j)], lc + rc + mc];
    }
    return mergeSort(arr)[1];
}

export function problem44(): void {
    console.log("Problem 44:");
    if (countInversions([2,4,1,3,5]) !== 3)  throw new Error("assertion failed");
    if (countInversions([5,4,3,2,1]) !== 10) throw new Error("assertion failed");
    if (countInversions([1,2,3])     !== 0)  throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_44.rs
pub fn count_inversions(arr: &[i64]) -> i64 {
    fn merge_sort(a: Vec<i64>) -> (Vec<i64>, i64) {
        if a.len() <= 1 { return (a, 0); }
        let mid = a.len() / 2;
        let (left, lc) = merge_sort(a[..mid].to_vec());
        let (right, rc) = merge_sort(a[mid..].to_vec());
        let mut merged = Vec::with_capacity(a.len());
        let (mut i, mut j, mut mc) = (0, 0, 0i64);
        while i < left.len() && j < right.len() {
            if left[i] <= right[j] { merged.push(left[i]); i += 1; }
            else { merged.push(right[j]); mc += (left.len() - i) as i64; j += 1; }
        }
        merged.extend_from_slice(&left[i..]);
        merged.extend_from_slice(&right[j..]);
        (merged, lc + rc + mc)
    }
    merge_sort(arr.to_vec()).1
}

pub fn problem44() {
    println!("Problem 44:");
    assert_eq!(count_inversions(&[2,4,1,3,5]), 3);
    assert_eq!(count_inversions(&[5,4,3,2,1]), 10);
    assert_eq!(count_inversions(&[1,2,3]),     0);
    println!("All Tests Passed");
}
